source/agora/common/Hash.d

```diff
@@ -289,3 +289,182 @@ nothrow @nogc @safe unittest
     auto hash_2 = hashMulti(420, "bpfk", S('a', 1, 'b', 2, 'c', 3));
     assert(hash_1 == hash_2);
 }
+
+/*******************************************************************************
+
+    Helper struct to manage a range of pre-images
+
+    When dealing with pre-images, the only unrecoverable data is the seed.
+    Everything else is, by definition, derived from the seed.
+    As a result, we can compress the data as much as we want,
+    to reduce memory usage. However, this comes at the cost of more computation.
+
+    For example, if we have cycles of 10 pre-images, we can first compute the
+    seed (H0), then reveal our initial commitment (H9). This will cost us
+    9 hash operations. Then on the next reveal, in order to get H8, we need to
+    perform 8 hash operations, then 7 for H7, etc...
+
+    If we want to improve this, we can store intermediate results:
+    when generating H9, we do perform 9 hash operations, but store H5.
+    Hence, when generating H8 and H7, we only need to perform
+    3 and 2 hash operations, respectively, and will never need to perform more
+    than 4 operations (save the initial commitment).
+
+    This structure is a mean to generalize this approach, with arbitrary count
+    and arbitrary sample size (interval, or distance, between saved pre-images).
+
+*******************************************************************************/
+
+public struct PreImageCache
+{
+    @safe nothrow:
+
+    /// Store the actual preimages
+    private Hash[] data;
+
+    /// Interval between two preimages in `data`
+    private const ulong interval;
+
+    /// Construct an instance using already existing data
+    public this (inout(Hash)[] data_, ulong sample_size) inout @nogc pure
+    {
+        this.data = data_;
+        this.interval = sample_size;
+    }
+
+    /***************************************************************************
+
+        Construct an instance and allocate memory
+
+        This takes a `count` of pre-image and a sample size, or distance.
+        For example, if one wish to store 1000 pre-images in this cache,
+        and perform no more than 5 hash operations each time,
+        the `sample_size` should be `5` and the `count` should be `200`
+        (1000 / 5).
```

Comment:
> if one wish to store 1000 pre-images in this cache

The wording threw me off a bit. I think you meant if one wishes to _use_ 1000 preimages. It only physically stores `count` number of images.

Reply:
"represent a range of 100 pre-images with this cache" ?

Reply:
Good!